[query] Use CRDD[Long] instead of CRDD[RegionValue] in most places

johnc1231 opened the pull request:
So this PR is mostly Patrick ripping out `RegionValues` in a lot of places in favor of just using a `Long`. I think this is useful because it simplifies the memory management situation by ensuring that the only regions to worry about at a particular `TableIR.execute` are the one on the `RegionContext` and any regions created in that `execute`. 

One exception to this is that currently we have methods `toCRDDRegionValue` and `toCRDDPtr` to switch back to a `RegionValue` based CRDD for compatibility with all the join stuff (really, anything that uses `OrderedRVIterator`).

In addition, this PR goes through and tries to systematically fix places where we were not manging regions correctly. This makes it somewhat subtle and hard to review. Some things I'd focus on are: 

- Anywhere a boundary was removed.
- `TableKeyByAndAggregate` and `TableAggregateByKey` (noted below. I haven't done anything aggregator related prior to this, don't fully understand them)
- Spot check of any of the places I listed under "Addressed" below.
- The implementations of `toCRDDRegionValue` and `toCRDDPtr`. 

Addressed:

```
- TableFilter
- TableSubset (just the parent of head and tail)
- TableHead (via rvd.head)
- TableTail (via rvd.tail)
- TableExplode
- LinearRegression
```

No change needed:

```
- TableMapGlobals
- TableRange
- TableLiteral
- TableOrderBy
- TableDistinct
- TableKeyBy
- TableRename
- TableUnion
- TableParallelize
- TableRead
```

Didn't / Minimally Changed, but wasn't really sure about:

```
- TableAggregateByKey
- TableKeyByAndAggregate
- TableMapRows (specifically, the bit about computing `scanPartitionAggs)
```

cc @patrick-schultz @cseed

johnc1231 commented:
Agh, failing one of my memory tests. Will dig in to it

tpoterba commented:
let me know when this is ready for review?

tpoterba commented:
is this ready for review?

johnc1231 commented:
I'm just waiting for benchmarks to run, but otherwise I think so

johnc1231 commented:
`large_range_matrix_table_sum()` failed in the benchmarks, looking into that. When I ran it locally, seemed to be allocating more memory than I would think, so there's probably a leak there. Otherwise, I think this is safe to review while I track this one down (and maybe you'll catch the cause of this)

```
2020-03-26 12:41:14 root: INFO: RegionPool: REPORT_THRESHOLD: 16.0M allocated (792.0K blocks / 15.3M chunks), thread 70: Executor task launch worker for task 10
```

tpoterba commented:
can you rebase? This is ready to go, right?

johnc1231 commented on hail/python/test/hail/table/test_table.py:
```diff
@@ -1388,7 +1388,7 @@ def test_write_table_containing_ndarray():
 
 def test_group_within_partitions():
-    t = hl.utils.range_table(10).naive_coalesce(2)
+    t = hl.utils.range_table(10).repartition(2)
```
Remember how Konrad complained that sometimes `naive_coalesce` gives you a different number of partitions than you ask for? That was happening here. If I don't know how many partitions I'm going to get for sure, the rest of this test doesn't work out. Patrick has encountered this as well, unclear why it happens (sometimes it happens locally but I haven't observed it during testing). Long story short, this is defending against an intermittent bug I don't understand

tpoterba replied:
ah, yes, of course.

![image](https://user-images.githubusercontent.com/10562794/78401771-792aa100-75c7-11ea-9ce8-2f41b312e184.png)

johnc1231 commented:
I believe I've addressed your comments, the benchmarks against 0.2.34 look normal, somehow there are no conflicts at this exact moment. Once #8474 goes in let's get this in.

tpoterba commented on hail/src/main/scala/is/hail/annotations/RegionValue.scala:
```diff
@@ -18,13 +18,9 @@ object RegionValue {
-      rv.setOffset(bad.regionValueFromBytes(r, bytes))
-      rv
-    }
+    byteses.map(bad.regionValueFromBytes(r, _))
```
> byteses

![image](https://user-images.githubusercontent.com/10562794/78592928-ce5af280-7813-11ea-9933-47bd2b654bcc.png)
